postgres: provide displayable format for messages.

cpakulski opened the pull request:
Commit Message:
Postgres messages are binary structures. They used to be sent to logger by treating them as `string` which very often resulted in log message containing weird characters. If log was sent to stdout, terminal went out of normal state and required a reset. 
This PR adds `toString` method which converts messages to displayable format. 

Risk Level: Low

Testing: Unit tests.

Docs Changes: No

Release Notes: Yes. Added description to Minor Changes section.

Fixes #11905

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+namespace NetworkFilters {
+namespace PostgresProxy {
+
+/*
```
Nit: `/**`

cpakulski replied:
Done.

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+  std::vector<uint8_t> value_;
+};
+
+// Class represents the structure consisting of 4 bytes of length
```
s/Class/VarByteN?

cpakulski replied:
Done.

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+// Class represents the structure consisting of 4 bytes of length
+// indicating how many bytes follow.
+// In Postgres documentation it is described as:
+// Int32
+// The length of the function result value, in bytes (this count does not include itself). Can be
+// zero. As a special case, -1 indicates a NULL function result. No value bytes follow in the NULL
+// case.
+//
+// ByteN
```
These should be bulleted points?

cpakulski replied:
Changed entire comment to be more descriptive.

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+// case.
+//
+// ByteN
+// The value of the function result, in the format indicated by the associated format code. n is the
+// above length.
```
Sorry, what is "n is the above length"?

cpakulski replied:
Changed entire comment for VarByteN.

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+    pos += sizeof(int32_t);
+    left -= sizeof(int32_t);
+    if (len_ < 1) {
+      // Nothing follows.
```
Can we expand it a bit more, or we can actually remove this comment.

cpakulski replied:
Expanded comment.

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+
+  std::string toString() const {
+    char buf[16];
+    sprintf(buf, "[(%d bytes):", len_);
```
This is a matter of preference I believe we can use `fmt::sprintf` here, the same as other places.

cpakulski replied:
Thanks for suggesting this. I changed all instances of `sprintf` to `fmt::format` and it drastically simplified the code.

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+// above length.
+class VarByteN {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
```
Not sure, but should this read be an explicit constructor of `VarByteN`?

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+// Size of integer types is fixed and depends on the type of integer.
+template <typename T> class Int {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
```
I see the pattern of having `read` method to initialize classes in this file. Do you think making it an explicit constructor makes sense?

cpakulski replied:
I prefer to leave it as it is. There are few reasons:
- `read` may fail and return an error. Returning errors in methods is easier than from constructors.
- variadic template `Sequence` chains `read`s. Implementing similar logic using constructors would be way harder .

dio replied:
Thanks!

jmarantz replied:
add doxygen comment for this method, maybe descriibing what happens when there's an error?

jmarantz replied:
thanks!

dio commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,330 @@
+};
+
+// Interface to Postgres message class.
+class MessageI {
```
Can we just spell this out as `MessageInterface`?

cpakulski replied:
Changed it to simply `Message` and marked methods as `PURE`. Most of base classes in Envoy which define API use the same approach.

dio reviewed (commented):
Flushing some comments. And I think it will be very valuable to get @jmarantz's review here.

fabriziomello commented on source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:
```diff
@@ -241,16 +254,26 @@ bool DecoderImpl::onData(Buffer::Instance& data, bool frontend) {
             std::get<0>(msg.get()));
-  ENVOY_LOG(debug, "({}) length = {}", msg_processor.direction_, getMessageLength());
-  ENVOY_LOG(debug, "({}) message = {}", msg_processor.direction_, getMessage());
+  ENVOY_LOG(debug, "({}) length = {}", msg_processor.direction_, message_len_);
```
What about move this two ENVOY_LOG to inside the branch below?

cpakulski replied:
Actually ENVOY_LOG is defined exactly in the same fashion: it checks the current logging level and logs only when logging level allows for logging. However, it only works when expression is included in ENVOY_LOG macro. Below (inside the branch) message is parsed outside of ENVOY_LOG macro and would be executed regardless of logging level.
https://github.com/envoyproxy/envoy/blob/master/source/common/common/logger.h#L344-L353

jmarantz replied:
WDYT about factoring lines 270-279 into a helper method, drop the if, and then you can simply put
```
ENVOY_LOG(debug, "({}) message = {}", msg_processor.direction_, genDebugMessage());
```
inline with the above logs. That way you don't repeat the level-filtering done in the macro.

fabriziomello commented on source/extensions/filters/network/postgres_proxy/postgres_decoder.h:
```diff
@@ -91,7 +89,8 @@ class DecoderImpl : public Decoder, Logger::Loggable<Logger::Id::filter> {
   // second - vector of Decoder's methods which are invoked when the message
   // is processed.
-  using MsgProcessor = std::pair<std::string, std::vector<MsgAction>>;
+  using MsgProcessor =
```
I never liked too much of the abbreviated version of `Message` word. What about rename all `Msg*` to `Message*`?

cpakulski replied:
I can change that for consistency and use `Message`.

cpakulski commented:
@jmarantz Would you mind reviewing this PR, as @dio suggested [here](https://github.com/envoyproxy/envoy/pull/13029#pullrequestreview-486529929), please? Thank you!

dio commented on source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:
```diff
@@ -241,16 +258,26 @@ bool DecoderImpl::onData(Buffer::Instance& data, bool frontend) {
+    std::string message = "Unrecognized";
+    if (f != nullptr) {
+      auto msgParser = f();
+      msgParser->read(data, message_len_ - 4);
```
Mind to put a comment on `- 4` here?

cpakulski replied:
Added the comment explaining why 4 bytes have to be subtracted here.

dio reviewed (commented):
Looks good. The test is fantastic. Thanks

dio commented:
/assign @jmarantz

for non-Tetrate review.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:
```diff
@@ -197,13 +217,10 @@ bool DecoderImpl::parseMessage(Buffer::Instance& data) {
   data.drain(startup_ ? 4 : 5); // Length plus optional 1st byte.
 
-  auto bytesToRead = length - 4;
+  auto bytesToRead = message_len_ - 4;
```
existing state, but per style guide can you use an explicit type here rather than `auto`, and also spell this `bytes_to_read`.

cpakulski replied:
Agreed. I corrected all `auto` types to explicit ones when calculations are done to initialize the variable.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,319 @@
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First find the terminating zero.
+    char zero = 0;
+    auto index = data.search(&zero, 1, pos);
```
explicit type

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,319 @@
+    }
+
+    // Reserve that many bytes in the string.
+    auto size = index - pos;
```
ditto. I'll give you a reason for this case, in addition to the style guide.

I'm always worried about unsigned arithmetic and overflow behavior  and I can't tell from context what the exact types are. So I can't simulate in my mind what the behavior will be.

cpakulski replied:
Agree. Corrected.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,319 @@
+    for (const auto& i : value_) {
+      std::string buf;
+      if (first) {
+        buf = fmt::format("{}", i);
```
no need to call format() here; you can just do absl::StrAppend(&out, i)

cpakulski replied:
Thanks for pointing it out. I thought that absl::StrAppend takes only strings as arguments.

jmarantz reviewed (commented):
flushing coments

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:
```diff
@@ -271,6 +271,9 @@ bool DecoderImpl::onData(Buffer::Instance& data, bool frontend) {
+      // message_len_ specifies total message length including 4 bytes long
+      // "length" field. The length of message body is total length minus size
+      // of "length" field (4 bytes).
       msgParser->read(data, message_len_ - 4);
```
s/message_len_ - 4/bytes_to_read/ ?

cpakulski replied:
Introduced new variable to skip carrying "-4".

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,352 @@
+
+  std::string toString() const { return fmt::format(getFormat(), value_); }
+
+  constexpr const char* getFormat() const { return "[{}]"; }
```
is there any benefit to having getFormat() as a function? In this case I think it might be no less readable, and a bit shorter, to just inline the string literal at the single call site.

cpakulski replied:
Thanks for catching it. It was leftover from parameterized test. Not needed anymore after moving to fmt::format.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,352 @@
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First find the terminating zero.
+    char zero = 0;
```
const, for both `zero` and `index`, and also `size` below.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,352 @@
+using Byte1 = Int<char>;
+
+// String type requires byte with zero value to indicate end of string.
+class String {
```
to me, all the implementations in this file (except the template one) seem like they should not be inlined into the header.

That means they need to be recompiled every time this .h is included, and will bloat debug info, etc.

Suggest having these just be declarations and putting the impls in a .cc

cpakulski replied:
Done. Moved all non-template methods to cc file.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,352 @@
+        first = false;
+      } else {
+        std::string buf;
+        buf = fmt::format(" {}", i);
```
no need for format call.  Actually I think you can just replace this loop with `absl::StrAppend(&out, absl::StrJoin(value_, " "))`

cpakulski replied:
Thanks for this suggestion. It drastically simplified the logic (and reduced the loop with dreaded "first" to just single line!)

cpakulski commented:
@jmarantz Thanks for reviewing. I addressed all your comments. Two CI tests are failing, not only in this PR but in few other PRs. It must be something I inherited during rebase from master.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_decoder.h:
```diff
@@ -87,20 +85,23 @@ class DecoderImpl : public Decoder, Logger::Loggable<Logger::Id::filter> {
   // is processed.
-  using MsgProcessor = std::pair<std::string, std::vector<MsgAction>>;
+  using MessageProcessor =
+      std::tuple<std::string, std::function<std::unique_ptr<Message>()>, std::vector<MsgAction>>;
```
can you make a separate `using` nickname for the function type, and then also use that name rather than `auto` in `genDebugMessage()` and possibly other places? I think that would make it easier to understand when reading.

cpakulski replied:
Added `MsgBodyReader` type.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,270 @@
+    pos += sizeof(uint16_t);
+    left -= sizeof(uint16_t);
+    if (num != 0) {
+      for (auto i = 0; i < num; i++) {
```
nit: s/auto/uint16_t/

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,270 @@
+    std::string out = fmt::format("[Array of {}:{{", value_.size());
+
+    for (const auto& i : value_) {
+      absl::StrAppend(&out, i->toString());
```
i'm assuming this outputs like you want without interposing a delimiter? Maybe just add a comment?

cpakulski replied:
Added comment.

jmarantz commented on source/extensions/filters/network/postgres_proxy/postgres_message.h:
```diff
@@ -0,0 +1,270 @@
+    std::string out;
+
+    for (const auto& i : value_) {
+      absl::StrAppend(&out, i->toString());
```
ditto -- comment why we don't need a delimiter between the stringified instances.

jmarantz reviewed (commented):
Looks great generally. A few minor nit-type comments.

In parallel to those being addressed I can pass this to @envoyproxy/senior-maintainers

jmarantz reviewed (approved):
@zuercher for senior maintainer approval

cpakulski commented:
@envoyproxy/senior-maintainers Can someone merge this PR, please? Thank you.
